p2panda-auth/src/group/mod.rs

```diff
@@ -0,0 +1,821 @@
+/// should be applied. The other is a special message which can be used to "undo" a message which
+/// has been applied to the group in the past.
+#[derive(Clone, Debug)]
+pub enum GroupControlMessage<ID, OP> {
+    /// An action to apply to the group state.
+    GroupAction {
+        group_id: ID,
+        action: GroupAction<ID>,
+    },
+
+    /// A revoke message can be published in order to explicitly invalidate other messages already
+    /// included in a group graph. This action is agnostic to any, probably more nuanced,
+    /// resolving logic which reacts to group actions.
+    ///
+    /// TODO: revoking messages is not implemented yet. I'm still considering if it is required in
+    /// or initial group implementation or something that can come later. There are distinct
+    /// benefits to revoking messages, over "just" making sure to resolve concurrent group action
+    /// conflicts (for example with "strong removal") strategy. By issuing a revoke message
+    /// revoking the message which first added a member into the group, it's possible to
+    /// completely erase that member from the group history. There can be an implicit "seniority"
+    /// rule in play, where it's only possible for an admin to revoke messages that they
+    /// published, or from when they were in the group.
+    Revoke { group_id: ID, id: OP },
+}
+
+impl<ID, OP> GroupControlMessage<ID, OP>
+where
+    ID: Copy,
+{
+    /// Returns true if this is a create control message.
+    pub fn is_create(&self) -> bool {
+        if let GroupControlMessage::GroupAction {
+            action: GroupAction::Create { .. },
+            ..
+        } = self
+        {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Id of the group this message should be applied to.
+    pub fn group_id(&self) -> ID {
+        match self {
+            GroupControlMessage::Revoke { group_id, .. } => *group_id,
+            GroupControlMessage::GroupAction { group_id, .. } => *group_id,
+        }
+    }
+}
+
+/// The inner state object for a single group.
+///
+/// This object has been separated from the GroupState so that it can be easily serialized and
+/// deserialized into a persistent store.
+///
+/// TODO: We may make custom implementations of serde traits where we "inject" the additional
+/// state required for GroupStore. In that case we don't need this separation and the inner and
+/// outer states could be combined again.  
+///
+/// TODO: One other reason these are separate structs is because I got into "recursive generic
+/// parameter" hell when trying to define a generic type (GroupStoreState) which returned another
+/// type (GroupState) which itself contained the container generic type (GroupStoreState)....
+/// Returning an "inner" state from the store (which doesn't know about the store) got around this
+/// issue.
+#[derive(Clone, Debug)]
+pub struct GroupStateInner<ID, OP, MSG>
```

Comment:
The handling of `GroupStateInner` is rather confusing and makes implementing the APIs hard, would be good to get rid of it or at least not surface it to the outside